calico/felix/masq.py

```diff
@@ -0,0 +1,86 @@
+# Copyright (c) Metaswitch Networks 2015. All rights reserved.
+
+import logging
+from calico.felix.actor import Actor, actor_message
+from calico.felix.futils import IPV4, IPV6
+from calico.felix.ipsets import Ipset, FELIX_PFX
+
+_log = logging.getLogger(__name__)
+
+
+ALL_POOLS_SET_NAME = FELIX_PFX + "all-ipam-pools"
+MASQ_POOLS_SET_NAME = FELIX_PFX + "masq-ipam-pools"
+
+
+class MasqueradeManager(Actor):
+    def __init__(self, ip_type, iptables_mgr):
+        super(MasqueradeManager, self).__init__(qualifier=str(ip_type))
+        assert ip_type in (IPV4, IPV6)
+        assert iptables_mgr.table == "nat"
+        self.ip_type = ip_type
+        self.pools_by_id = {}
+        self._iptables_mgr = iptables_mgr
+        ip_family = "inet" if ip_type == IPV4 else "inet6"
+        self._all_pools_ipset = Ipset(ALL_POOLS_SET_NAME,
+                                      ALL_POOLS_SET_NAME + "-tmp",
+                                      ip_family,
+                                      "hash:net")
+        self._masq_pools_ipset = Ipset(MASQ_POOLS_SET_NAME,
+                                       MASQ_POOLS_SET_NAME + "-tmp",
+                                       ip_family,
+                                       "hash:net")
+        self._dirty = False
+        self._install_rules(async=True)
+
+    @actor_message()
+    def _install_rules(self):
+        # Can't program a rule before we create a ipset but we won't know what
+        # to put in the ipset until we hear about it.
+        self._all_pools_ipset.ensure_exists()
+        self._masq_pools_ipset.ensure_exists()
+        # Enable masquerading for traffic coming from pools that have it
+        # enabled only when the traffic is heading to an IP that isn't in any
+        # Calico-owned pool.  (We assume that NAT is not required for
+        # Calico-owned IPs.)
+        self._iptables_mgr.ensure_rule_inserted(
+            "POSTROUTING "
+            "--match set --match-set %s src "
+            "--match set ! --match-set %s dst "
+            "--jump MASQUERADE" % (MASQ_POOLS_SET_NAME,
+                                   ALL_POOLS_SET_NAME),
+            async=True
+        )
```

Comment:
This rule will have a data plane performance impact in the mainline case when we aren't using this feature.  Is this acceptable?  We at least need to retest it (or change the spec for this PR so the rule isn't created).